### PR TITLE
Kratika - Solution

### PR DESCRIPTION
**Pull request by Kratika0907**

### CODE REVIEW

**Comment by ankushdharkar** [comment]
@Kratika0907 Updated the API format details in https://github.com/Real-Software-Dev-Squad/Max-Movies-In-Years/blob/master/README.md, please check and submit again in the same PR

**Comment by Kratika0907** [comment]
max-movies-in-years solution

File: Max-Movies-In-Years.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+let res = [
```
Comment (ankushdharkar) [inline Max-Movies-In-Years.js]: This should be `const`

File: Max-Movies-In-Years.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    name: "Iron Man",
+    license: [2005, 2008, 2011, 2012]
+  },
+  {
+    name: "Life of Pi",
+    license: [2008, 2011, 2014]
+  }
+];
+function getMaximumYears(res) {
+  let licenseArray = res.reduce((acc, element) => {
```
Comment (ankushdharkar) [inline Max-Movies-In-Years.js]: This should be `const`

File: Max-Movies-In-Years.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    return acc.concat(element.license);
+  }, []);
+  let yearMap = new Map();
```
Comment (ankushdharkar) [inline Max-Movies-In-Years.js]: This should be `const`

**Review by ankushdharkar — CHANGES_REQUESTED** [review]
Please use `const` for values that will not be changing their references

File: Max-Movies-In-Years.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+const res = [
```
Comment (ankushdharkar) [inline Max-Movies-In-Years.js]: Let's remove this file. Always try to get rid of stray code that's not serving purpose

File: Max-Movies-In-Years-2.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
+function displayMap(map) {
```
Comment (ankushdharkar) [inline Max-Movies-In-Years-2.js]: Indentations are all off. Please format properly

File: README.md [hunk 6]
```diff
@@ -1 +1,7 @@
-# Max-Movies-In-Years
\ No newline at end of file
+# Max-Movies-In-Years
+
+Max-Movies-In-Years-2 =>  
+
+Time Complexity = O(nlogn) where n is the number of element in years array
```
Comment (ankushdharkar) [inline README.md]: Please add some more details on why

File: Max-Movies-In-Years-2.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    ]
+
+    function getMaximumYears2(arr) {
+
+        // get the years and sort to get lowest and highest year 
+        const years = arr.reduce((acc,element)=> {
+            return acc.concat(element.year)
+            },[]).sort((a,b)=>a-b)
+       
+       const minYear =years[0];
```
Comment (ankushdharkar) [inline Max-Movies-In-Years-2.js]: What happens if `years` has a length of 0?

File: Max-Movies-In-Years-2.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+       const maxYear = years[years.length-1];
+       // map year to movie 
+       let yearMap = new Map();
+       for (let i = minYear ; i<= maxYear ; ++i) {
+           yearMap.set(i,0);
+       }
```
Comment (ankushdharkar) [inline Max-Movies-In-Years-2.js]: Maybe move to a separate function for readability?

File: Max-Movies-In-Years-2.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+       
+       arr.forEach((element)=>{
+           for (let i = element.year[0] ; i<=element.year[1];++i) {
+                yearMap.set(i,(yearMap.get(i)+1));
+       }
```
Comment (ankushdharkar) [inline Max-Movies-In-Years-2.js]: You can check if the map already has it and then add, or set as 1

Less code = Fewer mistakes + More confidence

File: Max-Movies-In-Years-2.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    year:[2007,2011]
+    }
+    ]
+
+    function getMaximumYears2(arr) {
+
+        // get the years and sort to get lowest and highest year 
+        const years = arr.reduce((acc,element)=> {
+            return acc.concat(element.year)
+            },[]).sort((a,b)=>a-b)
```
Comment (ankushdharkar) [inline Max-Movies-In-Years-2.js]: Do you *need* to sort?

Comment (ankushdharkar) [reply Max-Movies-In-Years-2.js]: Is the sort comparator correct? 

```js
year = [ 2001, 2002 ]
```

What is `(a - b)`?

File: README.md [hunk 11]
```diff
@@ -1 +1,7 @@
...
+
+Space Complexity = O(n) => map and array
```
Comment (ankushdharkar) [inline README.md]: Is this correct? Please re-evaluate

File: .prettierrc [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+    "trailingComma": "es5",
+    "tabWidth": 4,
+    "semi": false,
+    "singleQuote": true
```
Comment (ankushdharkar) [inline .prettierrc:5]: Don't push these unnecessary files please

File: Max-Movies-In-Years-2.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        name: "kratika3",
+        year: [2013, 2014]
+    },
+    {
+        name: "kratika4",
+        year: [2007, 2011]
+    }
+]
+
```
Comment (ankushdharkar) [inline Max-Movies-In-Years-2.js]: Any reason why there is `-2` in the name? and a file with `-3`

Remove dead code please

**Comment by Kratika0907** [comment]
Created a new PR while adhering comments and also added the discussed solution for same.
PR : [https://github.com/Real-Software-Dev-Squad/Max-Movies-In-Years/pull/15](url)

**Pull request closed without merging**